### PR TITLE
Shaving TTFT and reducing E2E latency with torch compile

### PR DESCRIPTION
**Pull request by mattteochen**
## Motivation

Several eager kernels are launched on the host before each CUDA graph replay in decode. This PR moves the replay-preparation path toward `torch.compile` so those small setup ops can be fused, reducing host-side decode overhead and the total number of launched kernels.

https://github.com/sgl-project/sglang/pull/18558 follow up cc @Fridge003 @nvpohanh.

For now, we supported only the `nsa_backend.py` attention backend.

## Modifications

- Refactored decode replay preparation in `CudaGraphRunner` into a compilable helper and enabled it only for whitelisted decode/idle backends.
- Added a small `torch.compile` utility with localized Dynamo/Inductor config patching plus warmup before replay.
- Made input-buffer population compile-friendly by batching inplace copies and allowing the compiled path to skip `_foreach_copy_`.
- Updated `TRTLLMHAAttnBackend` decode metadata initialization/replay to reuse preallocated `cache_seqlens` / `cu_seqlens` buffers, avoid storage aliasing, and add page-bound checks needed by compilation.

## Speed Tests and Profiling

Run on `B200` + `nsa` attention | `zai-org/GLM-5-FP8`.

### torch profiler
Baseline `648us`:
Screenshot 2026-04-13 at 19 10 01

Compiled `471us`:
Screenshot 2026-04-13 at 19 09 51

1.37x speedup for `replay_prepare`.

### Bench serving
command server:
```bash
python3 -m sglang.launch_server --port 8080 --model-path zai-org/GLM-5-FP8 --trust-remote-code --kv-cache-dtype fp8_e4m3 --tensor-parallel-size 8 --data-parallel-size 1 --expert-parallel-size 1 --enable-dp-lm-head --max-running-requests 1 --cuda-graph-max-bs 128 --mem-fraction-static 0.7 --chunked-prefill-size 32768 --max-prefill-tokens 32768 --enable-flashinfer-allreduce-fusion --disable-radix-cache --quantization fp8 --nsa-decode-backend trtllm --nsa-prefill-backend trtllm --moe-runner-backend flashinfer_trtllm --stream-interval 30
```

command client:
```bash
python3 -m sglang.bench_serving \
  --backend sglang \
  --model zai-org/GLM-5-FP8 \
  --dataset-name random \
  --random-input-len 1000 \
  --random-output-len 8000 \
  --num-prompts 10 \
  --max-concurrency 1 \
  --request-rate inf
  ```

Baseline:
```bash
=========== Serving Benchmark Result ===========
Backend:                             sglang
Traffic request rate:                inf
Max request concurrency:             1
Successful requests:                 10
Benchmark duration (s):              612.01
Total input tokens:                  6101
Total input text tokens:             6101
Total generated tokens:              44462
Total generated tokens (retokenized): 44451
Request throughput (req/s):          0.02
Input token throughput (tok/s):      9.97
Output token throughput (tok/s):     72.65
Peak output token throughput (tok/s): 78.00
Peak concurrent requests:            2
Total token throughput (tok/s):      82.62
Concurrency:                         1.00
----------End-to-End Latency----------
Mean E2E Latency (ms):               61198.42
Median E2E Latency (ms):             65971.14
P90 E2E Latency (ms):                102254.18
P99 E2E Latency (ms):                107800.69
----------Time to First Token----------
Mean TTFT (ms):                      1033.10
Median TTFT (ms):                    184.57
P99 TTFT (ms):                       2617.66
-----Time per Output Token (excl. 1st token)-----
Mean TPOT (ms):                      13.41
Median TPOT (ms):                    13.52
P99 TPOT (ms):                       13.64
----------Inter-Token Latency----------
Mean ITL (ms):                       13.53
Median ITL (ms):                     13.68
P95 ITL (ms):                        13.85
P99 ITL (ms):                        13.87
Max ITL (ms):                        15.50
================================================

```

Compiled:
```bash
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 1         
Successful requests:                     10        
Benchmark duration (s):                  601.97    
Total input tokens:                      6101      
Total input text tokens:                 6101      
Total generated tokens:                  44462     
Total generated tokens (retokenized):    44450     
Request throughput (req/s):              0.02      
Input token throughput (tok/s):          10.14     
Output token throughput (tok/s):         73.86     
Peak output token throughput (tok/s):    78.00     
Peak concurrent requests:                2         
Total token throughput (tok/s):          84.00     
Concurrency:                             1.00      
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   60194.72  
Median E2E Latency (ms):                 64816.88  
P90 E2E Latency (ms):                    101774.92 
P99 E2E Latency (ms):                    105285.85 
---------------Time to First Token----------------
Mean TTFT (ms):                          176.26    
Median TTFT (ms):                        177.44    
P99 TTFT (ms):                           179.31    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          13.38     
Median TPOT (ms):                        13.49     
P99 TPOT (ms):                           13.61     
---------------Inter-Token Latency----------------
Mean ITL (ms):                           13.50     
Median ITL (ms):                         13.65     
P95 ITL (ms):                            13.81     
P99 ITL (ms):                            13.83     
Max ITL (ms):                            13.88     
==================================================
```

Relevant metrics:
- Median E2E +1.8%
- Median TTFT +4%
- Output token throughput +1.6%

This is quite a big model, so CPU time latency is generally well amortized by the larger GPU time during decode. This feature is expected to provide more significant speedups for smaller model inference.

## Accuracy
gpqa
```
Repeat: 8, mean: 0.864
Scores: ['0.859', '0.859', '0.864', '0.864', '0.859', '0.869', '0.859', '0.884']
```

## Future work
A natural follow-up work is to compile in a single graph also the `prepare_for_decode` step from `schedule_batch.py`

---
### CI States

Latest PR Test (Base): :x: [Run #25739686358](https://github.com/sgl-project/sglang/actions/runs/25739686358)
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 1]
```diff
@@ -637,6 +645,10 @@ def __init__(self, model_runner: ModelRunner):
 
         self.tbo_plugin = TboCudaGraphRunnerPlugin()
 
+        # Combo scheduling is needed to have good performance for replay-prepare.
+        self.compile_replay_prepare = hasattr(torch._inductor.config, "combo_kernels")
```
Comment (mattteochen) [inline python/sglang/srt/model_executor/cuda_graph_runner.py]: Can be removed after https://github.com/sgl-project/sglang/pull/21247

**Comment by Fridge003** [comment]
@mattteochen Please test the gpqa/aime25 accuracy benchmark as the instruction here
https://docs.sglang.io/basic_usage/deepseek_v32.html#accuracy-test-with-gpqa-diamond

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 2]
```diff
@@ -79,6 +80,11 @@
 from sglang.srt.utils.patch_torch import monkey_patch_torch_compile
 from sglang.srt.utils.torch_memory_saver_adapter import TorchMemorySaverAdapter
 
+try:
+    torch._dynamo.config.capture_scalar_outputs = True
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/cuda_graph_runner.py:84]: Add comment on why we are doing this

File: python/sglang/srt/environ.py [hunk 3]
```diff
@@ -361,6 +361,7 @@ class Envs:
 
     # Torch Compile
     SGLANG_ENABLE_TORCH_COMPILE = EnvBool(False)
+    SGLANG_TORCH_COMPILE_CPP_WRAPPER = EnvBool(True)
```
Comment (Fridge003) [inline python/sglang/srt/environ.py:364]: Will it be risky to set this flag as True by default

File: python/sglang/srt/layers/attention/base_attn_backend.py [hunk 4]
```diff
@@ -18,6 +18,9 @@
 class AttentionBackend(ABC):
     """The base class of attention backends"""
 
+    # Whether init_forward_metadata_replay_cuda_graph is torch.compile-safe.
+    supports_compiled_replay_prepare: bool = False
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/base_attn_backend.py:22]: Make it a function that return False by default

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 5]
```diff
@@ -79,6 +80,11 @@
 from sglang.srt.utils.patch_torch import monkey_patch_torch_compile
 from sglang.srt.utils.torch_memory_saver_adapter import TorchMemorySaverAdapter
 
+try:
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/cuda_graph_runner.py:83]: Please move it to some place it's really needed. Like before calling _populate_from_forward_batch_and_init_attn_backend_compiled.
Silently setting it as a global value here is risky.

File: python/sglang/srt/environ.py (hunk 3 above)
Comment (Fridge003) [inline python/sglang/srt/environ.py:364]: Add this environ in docs/references/environment_variables.md

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 6]
```diff
@@ -637,6 +646,10 @@ def __init__(self, model_runner: ModelRunner):
 
         self.tbo_plugin = TboCudaGraphRunnerPlugin()
 
+        # Combo scheduling is needed to have good performance for replay-prepare.
+        self.compile_replay_prepare = hasattr(torch._inductor.config, "combo_kernels")
+        self.compiled_replay_prepare = None
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/cuda_graph_runner.py:651]: self.compiled_replay_prepare is not used, we can delete it

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 7]
```diff
@@ -637,6 +646,10 @@ def __init__(self, model_runner: ModelRunner):
 
         self.tbo_plugin = TboCudaGraphRunnerPlugin()
 
+        # Combo scheduling is needed to have good performance for replay-prepare.
+        self.compile_replay_prepare = hasattr(torch._inductor.config, "combo_kernels")
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/cuda_graph_runner.py:650]: We can remove self.compile_replay_prepare  and move the logic to _can_compile_replay_prepare

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py [hunk 8]
```diff
@@ -98,6 +100,9 @@ def __init__(
         self.page_size = model_runner.page_size
         self.req_to_token = model_runner.req_to_token_pool.req_to_token
         self.device = model_runner.device
+        self.max_num_pages = (
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/trtllm_mha_backend.py:103]: Please split the changes in trtllm_mha_backend.py to another PR, since it needs some different tests from nsa_backend.py

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 9]
```diff
@@ -75,6 +76,40 @@
...
+    batch_size = seqlens_k.shape[0]
+    tile_scheduler_metadata = seqlens_k.new_empty(
+        (num_sms // h_k, 8), dtype=torch.int32
+    )
+    num_splits = seqlens_k.new_empty((batch_size + 1,), dtype=torch.int32)
+    return [tile_scheduler_metadata, num_splits]
+
+
+@register_fake_if_exists("sgl_kernel::get_mla_decoding_metadata_dense_fp8")
+def _fake_get_mla_decoding_metadata_dense_fp8(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa_backend.py:101]: We can delete this function since it's not used

File: python/sglang/srt/environ.py [hunk 10]
```diff
@@ -361,6 +361,7 @@ class Envs:
...
```
Comment (Fridge003) [inline python/sglang/srt/environ.py:365]: Do we need an environment variable that controls this prepare metadata  torch compile feature. 
I feel it's still risky to enable this feature by default

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 11]
```diff
@@ -827,6 +915,21 @@ def init_forward_metadata_capture_cuda_graph(
...
+            nsa_cu_seqlens_k = self.decode_cuda_graph_metadata["nsa_cu_seqlens_k"][
+                : bs + 1
+            ]
+            nsa_cu_seqlens_k.copy_(compute_cu_seqlens(nsa_cache_seqlens_int32))
+            nsa_cu_seqlens_q = self.get_device_int32_arange(len(nsa_cu_seqlens_k))
+
+            # Use separate buffer for real_page_table to avoid aliasing
+            # with page_table_1 (when page_size == 1, _transform_table_1_to_real
+            # returns the same tensor)
+            real_page_table_src = self._transform_table_1_to_real(page_table_1)
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa_backend.py:928]: These copy operations might become bottleneck on long context batches. Can you do some benchmark/profile on long context like 128k?

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 12]
```diff
@@ -768,6 +822,28 @@ def init_cuda_graph_state(self, max_bs: int, max_num_tokens: int):
                 dtype=torch.int32,
                 device=self.device,
             ),
+            # Separate buffer for real_page_table to avoid storage aliasing
+            # with page_table (which triggers merge_view_inputs in torch.compile)
+            "real_page_table": torch.zeros(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa_backend.py:827]: Creating buffer for real_page_table will cost a lot of GPU space. Is there any other way to avoid merge_view_inputs

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 13]
```diff
@@ -1042,51 +1153,44 @@ def init_forward_metadata_replay_cuda_graph(
...
         if self.real_page_size > 1:
             real_table = self._transform_table_1_to_real(page_indices)
             new_rows = real_table.shape[0]
             new_cols = real_table.shape[1]
             metadata.real_page_table[:new_rows, :new_cols].copy_(real_table)
         else:
-            assert metadata.real_page_table is metadata.page_table_1
+            # real_page_table uses a separate pre-allocated buffer (no aliasing
+            # with page_table_1), so copy page_indices explicitly
+            metadata.real_page_table[
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa_backend.py:1191]: Also, the copy here will cost a lot of time on device

File: python/sglang/srt/layers/attention/nsa_backend.py
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa_backend.py:806]: Can we seperate the code paths w/ and wo/ compilation.
So the original code path doesn't change a lot

**Review by Fridge003 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 12 above)
Comment (mattteochen) [reply python/sglang/srt/layers/attention/nsa_backend.py]: For the CUDA DSA path, I think this is bounded: page_size=64, so `real_page_table` is `[max_num_tokens, max_context_len / 64 + 1]`, not a full duplicate of `[max_num_tokens, max_context_len]`. The large-memory concern mainly applies when `page_size==1`, because `_transform_table_1_to_real()` returns `page_table` directly and avoiding aliasing requires a full duplicate. I can add a guard so compiled replay-prepare/this separate buffer is only used for NSA when `real_page_size > 1` (64), as I think the `merge_input_view` overhead is not that simple to WAR.

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 11 above)
Comment (mattteochen) [reply python/sglang/srt/layers/attention/nsa_backend.py]: Running a `bench_one_batch.py` with `python3 -m sglang.bench_one_batch   --model-path zai-org/GLM-5-FP8   --trust-remote-code   --kv-cache-dtype fp8_e4m3   --tensor-parallel-size 8   --mem-fraction-static 0.8   --enable-flashinfer-allreduce-fusion   --quantization fp8   --nsa-decode-backend trtllm   --nsa-prefill-backend trtllm   --moe-runner-backend flashinfer_trtllm --output-len 128000` test (which should be enough to expose the copy ops repeated 128k times) leads to:
```
# Baseline
Benchmark ...
Prefill. latency: 0.16589 s, throughput:   6172.64 token/s
Decode.  median latency: 0.01603 s, median throughput:     62.40 token/s
Total. latency: 2036.955 s, throughput:     63.34 token/s

# Compiled
Benchmark ...
Prefill. latency: 0.16601 s, throughput:   6168.39 token/s
Decode.  median latency: 0.01587 s, median throughput:     63.03 token/s
Total. latency: 2015.729 s, throughput:     64.01 token/s
```

Verified from torch profile captured at 128k // 2:
Baseline:
Screenshot 2026-04-29 at 22 31 41

Compiled:
Screenshot 2026-04-29 at 22 06 01

**Comment by mattteochen** [comment]
Running the eval, I discovered that I'm suffering from several dynamo recompilations. I'm fixing those.

**Comment by mattteochen** [comment]
Updated the PR with accuracy values

**Comment by mattteochen** [comment]
Hi @Fridge003, I've updated the PR. Could you take a look again, please?

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 12 above)
Comment (mattteochen) [reply python/sglang/srt/layers/attention/nsa_backend.py]: I've guarded on page size.

**Comment by nvpohanh** [comment]
@mattteochen Could you fix the conflicts?
